pkg/stripe/analytics_telemetry.go

```diff
@@ -0,0 +1,169 @@
+//
+
+// TelemetryContextKey is the key for the telemetry context
+type TelemetryContextKey struct{}
+
+// TelemetryClientKey is the key for the telemetry client
+type TelemetryClientKey struct{}
+
+// DefaultTelemetryEndpoint is the default URL for the telemetry destination
+const DefaultTelemetryEndpoint = "https://r.stripe.com/0"
+
+// CLIAnalyticsEventContext is the structure that holds telemetry data context that is ultimately sent to the Stripe Analytics Service.
+type CLIAnalyticsEventContext struct {
+	InvocationID      string `url:"invocation_id"`      // The invocation id is unique to each context object and represents all events coming from one command / gRPC method call
+	UserAgent         string `url:"user_agent"`         // the application that is used to create this request
+	CommandPath       string `url:"command_path"`       // the command or gRPC method that initiated this request
+	Merchant          string `url:"merchant"`           // the merchant ID: ex. acct_xxxx
+	CLIVersion        string `url:"cli_version"`        // the version of the CLI
+	OS                string `url:"os"`                 // the OS of the system
+	GeneratedResource bool   `url:"generated_resource"` // whether or not this was a generated resource
+}
+
+// TelemetryClient is an interface that can send two types of events: an API request, and just general events.
+type TelemetryClient interface {
+	SendAPIRequestEvent(ctx context.Context, requestID string, livemode bool) (*http.Response, error)
+	SendEvent(ctx context.Context, eventName string, eventValue string) (*http.Response, error)
+}
+
+// AnalyticsTelemetryClient sends event information to r.stripe.com
+type AnalyticsTelemetryClient struct {
+	BaseURL    *url.URL
+	WG         *sync.WaitGroup
+	HTTPClient *http.Client
+}
+
+//
+// Public functions
+//
+
+// InitContext initializes an instance of CLIAnalyticsEventContext
+func InitContext() *CLIAnalyticsEventContext {
+	return &CLIAnalyticsEventContext{
+		InvocationID: uuid.NewString(),
+		UserAgent:    useragent.GetEncodedUserAgent(),
+		CLIVersion:   version.Version,
+		OS:           runtime.GOOS,
+	}
+}
+
+// SetCobraCommandContext sets the telemetry values for the command being executed.
+func (e *CLIAnalyticsEventContext) SetCobraCommandContext(cmd *cobra.Command) {
+	e.CommandPath = cmd.CommandPath()
+	e.GeneratedResource = false
+
+	for _, value := range cmd.Annotations {
+		// Generated commands have an annotation called "operation", we can
+		// search for that to let us know it's generated
+		if value == "operation" {
+			e.GeneratedResource = true
+		}
+	}
+}
+
+// SetMerchant sets the merchant on the CLIAnalyticsEventContext object
+func (e *CLIAnalyticsEventContext) SetMerchant(merchant string) {
+	e.Merchant = merchant
+}
+
+// SendAPIRequestEvent is a special function for API requests
+func (a *AnalyticsTelemetryClient) SendAPIRequestEvent(ctx context.Context, requestID string, livemode bool) (*http.Response, error) {
+	a.WG.Add(1)
```

Comment:
Could this be moved into `sendData` so it's not necessary to remember to do this in all of the public send methods?

Reply:
Great question! I actually had it in sendData initially but was worried that we could still run into race conditions where the program ends before we get through this setup portion of the public send methods, so figured it's safer to have it as early as possible. Not sure if that is a real concern in practice but if we want to make this more optional, what if I added an `add` and `defer` to `sendData` but also leave them in the public methods too? That way it's not necessary and just a precaution if the author of another method that uses `sendData` wants to add it?

Reply:
That makes sense to me!